CORE-13040 schema_registry: return max id on post schema endpoints

pgellert opened the pull request:
This PR:
1. Includes a refactoring of the schema registration logic, which is a follow up from https://github.com/redpanda-data/redpanda/pull/27079#discussion_r2259529075
2. Implements [CORE-13040](https://redpandadata.atlassian.net/browse/CORE-13040), which is to deterministically return the schema with the maximal schema id when looking up a schema by schema definition where there are multiple matches. This is necessary now that in import mode we allow posting schemas with identical definition at multiple schema ids. This is done for the two endpoints that perform lookups by schema definition: `POST /subjects/{subject}/versions/{version}` and `POST /subjects/{subject}`.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->


[CORE-13040]: https://redpandadata.atlassian.net/browse/CORE-13040?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

pgellert committed:
schema_registry: move `subject_version_entry` to types

pgellert committed:
schema_registry: add helpers for registration

Add helpers for the logic on the schema lookup/schema registration path.

pgellert committed:
schema_registry: merge `get_schema_version` into handler

pgellert committed:
schema_registry: reorder registration flow

Relocate mode checks and compatibility validation to the registration
section for better readability and logical flow.

No observable behaviour change is intended.

pgellert committed:
schema_registry: return max id in `POST /subjects/{subject}/versions/{version}`

When multiple registered schemas have identical definitions, return the one
with the highest schema ID instead of the first match found. This ensures
deterministic behavior and aligns with the reference implementation.

Note that the reference implementation returns the schema that was last
inserted from the topic in this case, so returning the max id is a
closest approximation in this edge case.

pgellert committed:
schema_registry: return max id in `POST /subjects/{subject}`

Apply the same maximum ID selection logic to the POST /subjects/{subject}
endpoint as the previous commit. When multiple registered schemas match
the submitted schema definition, return the one with the highest schema
ID.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--044b58e2-a6ca-4d91-ad45-81bb852e7874-->
<!--TEST RESULTS-->
<!--4fce92cf-a316-4847-857f-3e52948ae6cf-->
<!--TEST RESULTS-->
<!--58e8344f-969d-4df7-b53d-daeafeb72069-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70390">build#70390</a></summary>


| test_class                               | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogLimitPartitionBalancerTests | test_partition_balancer_with_limits | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70390#0198853e-436b-44bb-bf10-3c870a9573fa | FLAKY         | 20/21    | upstream reliability is '99.71264367816092'. current run reliability is '95.23809523809523'. drift is 4.47455 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70443">build#70443</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70443#01988928-df98-4853-b956-5e4bc4418fde | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70462">build#70462</a></summary>


| test_class               | test_method                 | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:-------------------------|:----------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations        | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70462#01988a16-be40-4a36-a880-a831cc5aae28 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| RapidTopicRecreateTest   | test_topic_rapid_recreation | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/70462#01988a18-d3dd-4947-bcbe-158d560801a0 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented:
> **consistently** return the schema with the maximal schema

(emphasis mine). what are the requirements around consistency (especially as it pertains to failure, fail-over, read-from-replica if that is a thing in SR), and how does the PR achieve the requirements?

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/e99c3fce360ea33b9d9957ff2f55fef7c48e1aee..67736ea1255d9a5faa70a191c6da186ca4c85a4b): fix clang tidy failure (use after move)

pgellert commented:
> > **consistently** return the schema with the maximal schema
> 
> (emphasis mine). what are the requirements around consistency (especially as it pertains to failure, fail-over, read-from-replica if that is a thing in SR), and how does the PR achieve the requirements?

@dotnwat Perhaps deterministically would be a better word here. It's possible to have multiple schema ids associated with the same schema definition, so if you look up a schema by definition, we have a choice to make on which schema id we want to return. This change is to deterministically choose the schema with the highest id.

There's no cross-node coordination happening here, so no concerns of consistency from the distributed systems sense. All of the schemas are stored in-memory on each broker locally.

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -185,18 +178,8 @@ sharded_store::get_schema_version(stored_schema schema) {
-          return res.value();
-      });
+    const auto versions = co_await get_subject_versions(
+      sub, include_deleted::no);
```
Probably worth fixing wither the comment or the code with regards to deleted.

pgellert replied:
Yeah the comment needed updating. I've pushed a commit to clear that up: https://github.com/redpanda-data/redpanda/pull/27167/commits/3eca74488369bd48031fd581864776696f7d0dee

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -406,6 +389,22 @@ sharded_store::get_schema_subject_versions(schema_id id) {
+            && res.assume_error().code() == error_code::subject_not_found) {
+              return {};
+          }
+          return res.value();
```
Wow, I was super strict when I wrote this!

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.h:
```diff
@@ -46,12 +46,6 @@ class sharded_store final : public schema_getter {
     ///\brief Construct a schema in the native format
     ss::future<valid_schema> make_valid_schema(subject_schema schema);
 
-    struct has_schema_result {
-        std::optional<schema_id> id;
-        std::optional<schema_version> version;
-    };
-    ss::future<has_schema_result> get_schema_version(stored_schema schema);
```
:fire:

BenPope commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -584,29 +574,13 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
-            return s_id_v.id == id;
-        });
+        auto v_it = std::ranges::find(
+          versions, *s_id, &subject_version_entry::id);
```
:fire: I was going to suggest this on the previous commit. We should probably just replace `absl` container algorithms globally. And I approve of the projection.

BenPope commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -621,16 +595,47 @@ post_subject_versions(server::request_t rq, server::reply_t rp) {
 
     schema_id schema_id{s_id.value_or(invalid_schema_id)};
     if (!matched) {
+        // Check if the request is appropriate for the mode
```
This ordering makes much more sense.

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -887,9 +887,10 @@ ss::future<bool> sharded_store::has_version(
+    auto reduce =
+      [](std::optional<schema_id> acc, std::optional<schema_id> s_id) {
+          return std::max(acc, s_id);
+      };
```
nitpick:
```suggestion
    auto reduce = [](auto a, auto b) { return std::max(a, b); };
```

BenPope commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -90,13 +90,11 @@ class store {
-        return s_it == _schemas.end() ? std::optional<schema_id>{}
-                                      : s_it->first;
+        // Iterate in decreasing order to return the maximal matching id
+        const auto& rev = std::ranges::reverse_view(_schemas);
```
nitpick:
```suggestion
        const auto& rev = std::views::reverse(_schemas);
```

pgellert committed:
schema_registry: clarify deleted version on register

Fix the outdated comment and rename the field for clarity.

On the schema registration/lookup endpoint we only want to consider
undeleted versions for lookup and compatibility checks.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/67736ea1255d9a5faa70a191c6da186ca4c85a4b..3eca74488369bd48031fd581864776696f7d0dee): address code review feedback

dotnwat commented:
> > > **consistently** return the schema with the maximal schema
> > 
> > 
> > (emphasis mine). what are the requirements around consistency (especially as it pertains to failure, fail-over, read-from-replica if that is a thing in SR), and how does the PR achieve the requirements?
> 
> @dotnwat Perhaps deterministically would be a better word here. It's possible to have multiple schema ids associated with the same schema definition, so if you look up a schema by definition, we have a choice to make on which schema id we want to return. This change is to deterministically choose the schema with the highest id.
> 
> There's no cross-node coordination happening here, so no concerns of consistency from the distributed systems sense. All of the schemas are stored in-memory on each broker locally.

Thanks @pgellert !
